crates/forge/src/run_tests/package.rs

```diff
@@ -0,0 +1,181 @@
+use super::{
+    resolve_config::resolve_config,
+    test_target::{run_test_target, TestTargetRunResult},
+};
+use crate::{
+    block_number_map::BlockNumberMap,
+    combine_configs::combine_configs,
+    pretty_printing,
+    scarb::{
+        config::{ForgeConfigFromScarb, ForkTarget},
+        load_test_artifacts,
+    },
+    shared_cache::FailedTestsCache,
+    test_filter::TestsFilter,
+    warn::{
+        warn_if_available_gas_used_with_incompatible_scarb_version,
+        warn_if_incompatible_rpc_version,
+    },
+    TestArgs,
+};
+use anyhow::Result;
+use camino::{Utf8Path, Utf8PathBuf};
+use cheatnet::runtime_extensions::forge_runtime_extension::contracts_data::ContractsData;
+use configuration::load_package_config;
+use forge_runner::{
+    forge_config::ForgeConfig,
+    package_tests::{
+        raw::TestCrateRaw, with_config::TestTargetWithConfig,
+        with_config_resolved::TestTargetWithResolvedConfig,
+    },
+    test_case_summary::AnyTestCaseSummary,
+    test_crate_summary::TestTargetSummary,
+};
+use scarb_api::get_contracts_artifacts_and_source_sierra_paths;
+use scarb_metadata::{Metadata, PackageMetadata};
+use std::{env, sync::Arc};
+
+pub struct RunFromCrateArgs {
+    pub test_targets: Vec<TestTargetWithConfig>,
+    pub tests_filter: TestsFilter,
+    pub forge_config: Arc<ForgeConfig>,
+    pub fork_targets: Vec<ForkTarget>,
+    pub package_name: String,
+}
+
+pub fn prepare_package(
+    package: PackageMetadata,
+    scarb_metadata: &Metadata,
+    args: &TestArgs,
+    cache_dir: Utf8PathBuf,
+    snforge_target_dir_path: &Utf8Path,
+    versioned_programs_dir: Utf8PathBuf,
+) -> Result<RunFromCrateArgs> {
+    env::set_current_dir(&package.root)?;
```

Comment:
This should be moved to the caller (so anything that messes with the state is easily visible) and then the function can be a method named `build` on `RunFromCrateArgs`